### PR TITLE
Mem address gen

### PR DESCRIPTION
**Pull request by stepeh**
This is an N-dimensional blocking address generator. It performs the same function as the 3D blocking address generator (and so can easily replace it), except that it also works for any number of dimensions (including 1, which means it can also replace the linear address generator).

A major advantage over the regular address generators is that you can specify multiple blocks to be read from memory, which would allow a MaxGenFD style application to run a whole timestep without a reset. There is another option to get it to repeat a set pattern many times, which means a MaxGenFD style application could actually run every timestep with a single action.

Basically this is what you would need if you wanted to do MaxGenFD (or any grid based computation) properly in any number of dimensions, but it could also be used for many other things.

Another advantage is that this doesn't require a reset, so it might be possible to use it in a networking application, although I can't think of a use case.

### CODE REVIEW

File: src/maxpower/lmem/addressgenerators/MultiDimensionalAddressGenerator.maxj [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+		flush.onTrigger(trigger);
+	}
+
+	//Avoid offsetting to before cycle 0.
+	private DFEVar safeOffset(DFEVar input, int offset) {
+		if (offset > 0) {
+			throw new RuntimeException("safeOffset is only for negative stream offsets.");
+		}
+		Params params = control.count.makeParams(MathUtils.bitsToRepresent(-offset+1))
+		                             .withMax(-offset+1)
```
Comment (chr1sj0nes) [inline src/maxpower/lmem/addressgenerators/MultiDimensionalAddressGenerator.maxj]: I think the +1 is unnecessary.

File: src/maxpower/lmem/addressgenerators/MultiDimensionalAddressGenerator.maxj [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+			}
+			DFEVar offset = chain.addCounter(numBursts,  128);
+			finalCmd &= offset + 128 >= numBursts;
+
+			enableOutput = constant.var(true);
+
+			DFEVar burstsLeft = numBursts - offset;
+			address  = startAddress + offset;
+			size     = burstsLeft < 128 ? burstsLeft.cast(sizeType) : 128;
+		} else {
```
Comment (chr1sj0nes) [inline src/maxpower/lmem/addressgenerators/MultiDimensionalAddressGenerator.maxj]: I find the amount of duplicated code here quite upsetting. Perhaps create a makeScalarOrRom() method?

File: test/maxpower/lmem/addressgenerators/MultiDimensionalAddressGeneratorTest.maxj [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+import static org.junit.Assert.assertTrue;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+import org.junit.Test;
+
+import com.maxeler.maxcompiler.v0.utils.MathUtils;
```
Comment (chr1sj0nes) [inline test/maxpower/lmem/addressgenerators/MultiDimensionalAddressGeneratorTest.maxj]: Probably better to use v2 MathUtils.

File: test/maxpower/lmem/addressgenerators/MultiDimensionalAddressGeneratorTest.maxj [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+import com.maxeler.maxcompiler.v2.kernelcompiler.stdlib.LMemCommandStream;
+import com.maxeler.maxcompiler.v2.managers.standard.SimulationManager;
+
+public class MultiDimensionalAddressGeneratorTest {
+
+	private class LAGTestData {
```
Comment (chr1sj0nes) [inline test/maxpower/lmem/addressgenerators/MultiDimensionalAddressGeneratorTest.maxj]: This is pure pedantry, but I assume the LAG stands for LinearAddressGenerator, but is now being used for all the tests.
